keps/sig-apps/0034-maxunavailable-for-statefulsets.md

```diff
@@ -0,0 +1,179 @@
+that requires diving into the complicacy of hashing and how replicasets are named. Over and above that, there are some issues with hash collisions which 
+further complicate the situation(I know they were solved). StatefulSet introduced ControllerRevisions in 1.7 which I believe are easier to think and reason 
+about. They are used by DaemonSet and StatefulSet for tracking revisions. It would be so much nicer if all the use cases of Deployments can be met and we 
+could track the revisions by ControllerRevisions.
+
+With this feature in place, when using StatefulSet with maxUnavailable >1, the user understands that this would not cause issues with their Stateful 
+Applications which have per pod state and identity while still providing all of the above written advantages.
+
+### Goals
+StatefulSet RollingUpdate strategy will contain an additional parameter called `maxUnavailable` to control how many Pods will be brought down at a time,
+during Rolling Update.
+
+### Non-Goals
+maxUnavailable is only implemeted to affect the Rolling Update of StatefulSet. Considering maxUnavailable for Pod Management Policy is beyond the purview
+of this KEP.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As a User of Kubernetes,  I can create a StatefulSet with RollingUpdate Strategy and specify maxUnavailable.
+
+### Implementation Details
+
+#### API Changes
+
+Following changes will be made to the Rolling Update Strategy for StatefulSet.
+
+```go
+// RollingUpdateStatefulSetStrategy is used to communicate parameter for RollingUpdateStatefulSetStrategyType.
+type RollingUpdateStatefulSetStrategy struct {
+        // Partition indicates the ordinal at which the StatefulSet should be
+        // partitioned.
+        // Default value is 0.
+        // +optional
+        Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
+
+	// The maximum number of pods that can be unavailable during the update.
+        // Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+        // Absolute number is calculated from percentage by rounding down.
+        // Defaults to 1.
+        // +optional
+        MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,2,opt,name=maxUnavailable"`
+	
+	...
+}
+```
+
+- By Default, if maxUnavailable is not specified, its value will be assumed to be 1 and StatefulSets will follow their old behavior.
+- if MaxUnavailable is specified, it cannot be greater than total number of replicas
+- If a partition is specified, maxUnavailable will only apply to all the pods which are staged by the partition. Lets say total replicas
+is 5 and partition is set to 2 and maxUnavailable is set to 2. If the image is changed in this scenario, pods with ordinal 4 and 3 will go
+down at the same time(because of maxUnavailable), once they are running and ready, pods with ordinal 2 will go down. Pods with ordinal 0
```

Comment:
> once they are running and ready, pods with ordinal 2 will go down

I think this may change to `once one of 3 or 4 is running and ready, pods with ordinal 2 will go down`.

Reply:
@FillZpp the ordering is still maintained. So until both 3 and 4 are both updated, it cannot proceed to 2. Its a different debate whether we want to support maxUnavilable for parallel pod management policy as well or not, in which case your assumption will be true. I dont see any drawbacks with support maxUnavailable with parallel pod management.  Looking for others to chime in on any gotchas, issues with doing that.

Reply:
@krmayankk  I'm not completely agree with you. 

If 2 can only update when both 3 and 4 are already updated, then it is not `maxUnavailable` strategy, it may be called `batch` or something. `maxUnavailable` means there are most 2 pods updating at one time, like `Sliding Window`.

So i think this maybe the right way:
- If `podManagementPolicy==Parallel`, then `
once one of 3 or 4 is running and ready, pod with ordinal 2 will go down`
- if `podManagementPolicy==OrderedReady` by default, then `once 4 is running and ready,  pod with ordinal 2 will go down`

There is no need to wait for both 3 and 4 are running and ready.

Reply:
@FillZpp i am open to suggestions here on what the right behavior should be. Your suggestion for orderedready seem to violate the ordering. Since 2 is going down before 3. Also its not clear what is the difference in your example between parallel and orderedready ?

My suggestion is still the following:-
- for podManagementPolicy==OrderedReady, for replicas=5 and partition=2, maxUnavailable =2,  4 and 3 will start updating at the same time, once both have finished updating, 2 will go down
- for podManagementPolicy==Parallel, i need to test if partition affects it or not. I thought partition doesnt change anything. But i think what you are suggesting makes sense . Basically any 2 can go down, for e.g. 2 and 3 and then in the end 4.

Reply:
@krmayankk I don't think since 4 is finished, then updating 3 and 2 together violates the ordering, it is not `2 going down before 3`. 

The ordering is 3 and 4 updating first, and once 4 is finished, you can update 2 immediately. So 2 maybe update together with 3, but it is not updating before 3. It's just like you update 4 and 3 at first.

For example, for `podManagementPolicy==OrderedReady && replicas==10 && partition=2 && maxUnavailable =2`, you think both 8&9 finished, then 6&7, then 4&5, then 2&3 ? I don't think this is 'maxUnavailable', we may call it 'batch' strategy.

What do you think about it? @kow3ns @janetkuo @kargakis

Reply:
What @FillZpp describes makes sense to me re how OrderedReady and MaxUnavailable interact. I can't think of any use case for such a combo though to be honest. Do you guys have any in mind?